src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs

```diff
@@ -250,13 +251,35 @@ private static Rect GetTransformedBounds(IVisual visual)
 
                 if ((containsPoint || !visual.ClipToBounds) && visual.VisualChildren.Count > 0)
                 {
-                    foreach (var child in visual.VisualChildren.SortByZIndex())
+                    var childrenCount = visual.VisualChildren.Count;
+                    
+                    if (childrenCount > 1)
                     {
-                        foreach (var result in HitTest(child, p, filter))
+                        var childrenArray = ArrayPool<IVisual>.Shared.Rent(childrenCount);
```

Comment:
1) ArrayPool can return an array larger than requested, so some elements will be null
2) try/copy/finally/return pattern is too bloated

Consider using `PooledList<T>` instead and probably adding a constructor that takes an existing IReadOnlyList<T> to copy contents from.

Reply:
I have a question about this. I already account for ArrayPool adding extra elements by iterating with for instead of foreach. However, looking at PooledList<T>, I see that creating it every time might also extensively allocate on it's own cos the object itself consists of multiple ints and pointers that will have to be allocated every time this is created. It will probably be beneficial when multiple children exist but I think that a more common use case is that IVisual has a few children in which case I think using ArrayPool directly actually significantly helps with performance.

Reply:
I have a question about this. I already account for ArrayPool adding extra elements by iterating with for instead of foreach. However, looking at PooledList, I see that creating it every time might also extensively allocate on it's own cos the object itself consists of multiple ints and pointers that will have to be allocated every time this is created. It will probably be beneficial when multiple children exist but I think that a more common use case is that IVisual has a few children in which case I think using ArrayPool directly actually significantly helps with performance.

Reply:
So is the list itself possibly getting cached somewhere?

Reply:
I think using `PooledList` here can cause extra allocations that will be hard to avoid. My primary worry is that with deep control hierarchies (`TreeView` in ControlCatalog for example) default pooling stops working completely and adds extra overhead.

This is caused by a limit of 8 pooled arrays per size (and per core) so when rendering and reaching lower depths pooled arrays will be allocated and discarded after use. We could investigate if we can rewrite current recursion into iteration so one shared render/hit test pool can be kept.

Reply:
I addressed these problems in the following manner:

I created two Renderer-wise PooledLists<(int, IVisual)> - one for hit testing, the other for rendering.

The reason why they hold int is to be able to not allocate for sorting and just use a pre-allocated sorter from the ImmediateRenderer (new).

The int is there to keep the original position in the parent's visualChildren and is used for stable sorting.

The pooled lists are used as sort of stacks where ranges are added and removed once the recursion branch is finished with them.

I also reworked the hit test to have separate implementations for the HitTestFIrst that was allocating way too much when it just needs to return the first element hit. For the HitTest that returns IEnumerable I actually reverted to the old implementation.

The measurement after this show 0 per frame allocations from the ImmediateRenderer.

Reply:
>However, looking at PooledList, I see that creating it every time might also extensively allocate on it's own cos the object itself consists of multiple ints and pointers that will have to be allocated every time this is created.

You can use ObjectPool to pool PooledList instances.

Reply:
I redesigned how I approach the whole idea so I don't think that such pooling is needed now. I use only two PooledList objects per lifetime of the ImmediateRenderer.